protoc-gen-swagger/genswagger/template.go

```diff
@@ -288,74 +279,101 @@ func renderMessagesAsDefinition(messages messageMap, d swaggerDefinitionsObject,
 		if opt := msg.GetOptions(); opt != nil && opt.MapEntry != nil && *opt.MapEntry {
 			continue
 		}
-		schema := swaggerSchemaObject{
-			schemaCore: schemaCore{
-				Type: "object",
-			},
+		d[fullyQualifiedNameToSwaggerName(msg.FQMN(), reg)] = renderMessageAsSchema(msg, reg, customRefs, nil)
+	}
+}
+
+func renderMessageAsSchema(msg *descriptor.Message, reg *descriptor.Registry, customRefs refMap, excludeFields []*descriptor.Field) swaggerSchemaObject {
```

Comment:
this was lifted and modified from the loop body in `renderMessagesAsDefinition` - I recommend reviewing with "Hide whitespace changes" set to avoid the indentation change obscuring the logic diff.